webapi/setaltsig.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/binding"
+)
+
+// Ensure that Node is satisfied by *rpc.DcrdRPC.
+var _ Node = (*rpc.DcrdRPC)(nil)
+
+// Node is satisfied by *rpc.DcrdRPC and retrieves data from the blockchain.
+type Node interface {
+	CanTicketVote(rawTx *dcrdtypes.TxRawResult, ticketHash string, netParams *chaincfg.Params) (bool, error)
+	GetRawTransaction(txHash string) (*dcrdtypes.TxRawResult, error)
+}
+
+// setAltSig is the handler for "POST /api/v3/setaltsig".
+func setAltSig(c *gin.Context) {
+
+	const funcName = "setAltSig"
+
+	// Get values which have been added to context by middleware.
+	dcrdClient := c.MustGet("DcrdClient").(Node)
+	reqBytes := c.MustGet("RequestBytes").([]byte)
+
+	if cfg.VspClosed {
+		sendError(errVspClosed, c)
+		return
+	}
+
+	var request SetAltSigRequest
+	if err := binding.JSON.BindBody(reqBytes, &request); err != nil {
+		log.Warnf("%s: Bad request (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+
+	altSigAddr, ticketHash := request.AltSigAddress, request.TicketHash
+
+	currentData, err := db.AltSigData(ticketHash)
+	if err != nil {
+		log.Errorf("%s: db.AltSigData (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if currentData != nil {
+		msg := "alternate signature data already exists"
+		log.Warnf("%s: %s (ticketHash=%s)", funcName, msg, ticketHash)
+		sendErrorWithMsg(msg, errBadRequest, c)
+		return
+
+	}
+
+	// Fail fast if the pubkey doesn't decode properly.
+	addr, err := stdaddr.DecodeAddressV0(altSigAddr, cfg.NetParams)
+	if err != nil {
+		log.Warnf("%s: Alt sig address cannot be decoded (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+	if _, ok := addr.(*stdaddr.AddressPubKeyHashEcdsaSecp256k1V0); !ok {
+		log.Warnf("%s: Alt sig address is unexpected type (clientIP=%s, type=%T)", funcName, c.ClientIP(), addr)
+		sendErrorWithMsg("wrong type for alternate signing address", errBadRequest, c)
+		return
+	}
+
+	// Get ticket details.
+	rawTicket, err := dcrdClient.GetRawTransaction(ticketHash)
+	if err != nil {
+		log.Errorf("%s: dcrd.GetRawTransaction for ticket failed (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+
+	// Ensure this ticket is eligible to vote at some point in the future.
+	canVote, err := dcrdClient.CanTicketVote(rawTicket, ticketHash, cfg.NetParams)
+	if err != nil {
+		log.Errorf("%s: dcrd.CanTicketVote error (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if !canVote {
+		log.Warnf("%s: unvotable ticket (clientIP=%s, ticketHash=%s)",
+			funcName, c.ClientIP(), ticketHash)
+		sendError(errTicketCannotVote, c)
+		return
+	}
+
+	sigStr := c.GetHeader("VSP-Client-Signature")
+
+	// Send success response to client.
+	res, resSig := sendJSONResponse(SetAltSigResponse{
```

Comment:
This could be a silly question (if this line means "queue" and not "send") but I'll take the beating for a chance to spot a problem.

Could there be an issue with sending success _before_ saving data to the db?

Reply:
This does return a response instantly, its not queued. This is a good catch actually, I overlooked it because we do it in other places, but in those cases we are only storing an audit trail. This case is actually storing some important data, and if this db call fails we should return an error to the client. I will fix this.